MoltenVK/MoltenVK/Commands/MVKCmdTransfer.mm

```diff
@@ -519,6 +519,8 @@ static inline MTLSize mvkClampMTLSize(MTLSize size, MTLOrigin origin, MTLSize ma
                     mtlStencilAttDesc.slice = mvkIBR.region.dstSubresource.baseArrayLayer + layIdx;
                 }
                 id<MTLRenderCommandEncoder> mtlRendEnc = [cmdEncoder->_mtlCmdBuffer renderCommandEncoderWithDescriptor: mtlRPD];
+				cmdEncoder->recordTimestamp(mtlRendEnc);
```

Comment:
That doesn't work because the sampling command is issued on a per-encoder basis. Encoder might end encoding before the timestamp command is requested, or the user might have not even started the encoder if the timestamp is send before doing any other encoder commands.

I tried other approaches, including creating temporary encoders, but that resulted in invalid timings. This only way to catch all the correct timings is to wrap all encoder calls with a recording of the time it started and ended.

If you wanted to optimize this you could do a 'lookahead' and avoid recording the timestamp if there are no `vkCmdWriteTimestamp` commands queued in the command buffer.

Reply:
> I don't see us using `sampleBufferAttachments`, if they exist. Could that be part of the inaccuracy? What does the timestamp represent if no `sampleBufferAttachment` has been configured?

I think that's for `MTLCounterSamplingPointAtStageBoundary`, which this change doesn't support--I assume deliberately so, since that would make this even more complex. The version on Catalina doesn't support that at all. So I assume it isn't necessary.

Reply:
Does setting `withBarrier: YES` help with the accuracy? According to the Vulkan spec:

> `vkCmdWriteTimestamp` latches the value of the timer *when all previous commands have completed executing* as far as the specified pipeline stage, and writes the timestamp value to memory. [emphasis added]

`withBarrier: YES` inserts a barrier that waits for the previous commands to finish, which is what we want here.

Reply:
You are right that `recordTimestamp()` doesn't handle the case for multiple draw calls within the same render pass. That is an oversight.

As I mentioned creating a temporary encoder did not work for me. I was seeing invalid times (nested timestamp samples reporting higher times than their 'parent' for example). But I have not tried with the barrier option, as the docs state it might negatively impact performance.